README.md
# Services

> Microsserviços essenciais para o Instabotx.

![version](https://img.shields.io/badge/version-0.1.0-blue)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/49ba3ec324d04f789f4650d07dd5434e)](https://www.codacy.com/gh/instabotix/services/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=instabotix/services&amp;utm_campaign=Badge_Grade)

## Serviços

- [`productFinder.js`](productFinder.js) fornece o ID dos produtos mostrados em cada mídia do Instagram de acordo com o 
  que foi cadastrado na plataforma da VTEX
- [`commentWatcher.js`](commentWatcher.js) verifica as notificações da conta do Instagra, selenciona apenas os 
  comentários e encaminha para o chatbot


CHANGELOG.md
# Changelog
Todas as mudanças notáveis neste projeto serão documentadas neste arquivo.

Este formato é baseado em [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
e este projeto segue [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Não publicado]

## [0.1.0] - 2021-04-02
### Adicionado
- Este arquivo CHANGELOG
- Arquivo [README](README.md) com informações iniciais do projeto
- [ROADMAP](ROADMAP.md) com as próximas funcionalidades
- Rotas `GET`, `POST`, `PUT` e `DELETE` para `mídias`
- Conexão com banco de dados

[Não publicado]: https://github.com/instabotix/services/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/instabotix/services/releases/tag/v0.1.0


app.js
// Documento principal da aplicação do bot
// Iniciando o servidor Express JS
const express = require('express');
const app = express();


// O cors permite compartilhar recursos entre diferentes endereços http
const cors = require('cors');
app.use(cors()); //definindo o cors como um middleware


// Obtendo as variáveis do ambiente
const PORT = process.env.PORT || 3000;
const MONGODB_URL = process.env.MONGODB_URL || 'mongodb://user:password@localhost/produtos';


// Configurando o banco de dados
const mongoose = require('mongoose');
mongoose.connect(MONGODB_URL, {
    useFindAndModify: false,
    useNewUrlParser: true,
    useUnifiedTopology: true
});

// Mensagem de erro
mongoose.connection.on('error', (err) => {
    console.log('Erro no banco de dados ' + err);
});

const produtoSchema = mongoose.Schema({
    media_id: String,
    produtos: [String]
}, {collection: 'produtos'});

const Produtos = mongoose.model('Produtos', produtoSchema);


// Habilitando para trabalhar com dados em JSON
app.use(express.json());


// Rotas
app.put('/produtos', (req, res, next) => {
    Produtos.create(req.body).then((produto) => {
        res.send(produto);
    }).catch(next);
});

app.get('/produtos/:media_id', (req, res, next) => {
    Produtos.findOne({media_id: req.params.media_id}).then((produto) => {
        res.send(produto);
    }).catch(next);
});

app.post('/produtos/:media_id', (req, res, next) => {
    Produtos.findOneAndUpdate({media_id: req.params.media_id}, req.body).then(() => {
        Produtos.findOne({media_id: req.params.media_id}).then((produto) => {
            res.send(produto);
        }).catch(next);
    }).catch(next);
});

app.delete('/produtos/:media_id', (req, res, next) => {
    const media_id = req.params.media_id;
    Produtos.findOneAndRemove({media_id: req.params.media_id}).then((produto) => {
        res.send(produto);
    }).catch(next);
});

app.use((err, req, res, next) => {
    res.status(422).send({erro: err.message});
});

app.listen(PORT);


commentWatcher.js
const Instagram = require('instagram-web-api');
const mongoose = require('mongoose');


// Obtendo as variáveis de ambiente
const MONGOBD_URL = process.env.MONGODB_URL;
const [ username, password ] = [process.env.INSTAGRAM_USERNAME, process.env.INSTAGRAM_PASSWORD];


// Configurando o banco de dados
mongoose.connect(MONGOBD_URL, {
    useFindAndModify: false,
    useNewUrlParser: true,
    useUnifiedTopology: true
});

mongoose.connection.on('error', (err) => {
    console.log('Erro no banco de dados ' + err);
});

// Esquema da coleção que registra os comentários que já foram encaminhados para o chatbot
const comentarioSchema = new mongoose.Schema({
    comentario: {type: String, unique: true}
}, {collection: 'comentarios'});

// Registra o modelo no banco de dados
const Comentarios = mongoose.model('Comentarios', comentarioSchema);


// Biblioteca da API não-oficial do Instagram
const client = new Instagram({username, password});

(async () => {
    await client.login();
    const atividade = await client.getActivity();
    const notificacoes = atividade.activity_feed.edge_web_activity_feed.edges;

    notificacoes.forEach((notificacao) => {
        // Seleciona apenas as notificação referentes a comentários
        if (notificacao.node.__typename === 'GraphCommentMediaStory') {
            const comentario_id = notificacao.node.id.split('_').slice(1).join('_');
            // Procura no banco de dados se o comentário já foi endereçado
            Comentarios.find({comentario: comentario_id}, (err, doc) => {
                if (doc.length === 0) {
                    // Seleciona os detelhes do comentário para enviar para o chatbot
                    Comentarios.create({comentario: comentario_id}, (err, doc) => {
                        let comentario = {
                            comentario_id: comentario_id,
                            timestamp: notificacao.node.timestamp,
                            usuario: {
                                id: notificacao.node.user.id,
                                usuario: notificacao.node.user.username,
                                nome_completo: notificacao.node.user.full_name
                            },
                            media_id:  notificacao.node.media.shortcode,
                            comentario: notificacao.node.text
                        }
                        console.log(comentario);
                    });
                }
            });
        }
    });
})();


productFinder.js
// Documento principal da aplicação do bot
// Iniciando o servidor Express JS
const express = require("express");
const app = express();


// O cors permite compartilhar recursos entre diferentes endereços http
const cors = require("cors");
app.use(cors()); //definindo o cors como um middleware


// Obtendo as variáveis do ambiente
const PORT = process.env.PORT || 3000;
const MONGODB_URL = process.env.MONGODB_URL || "mongodb://user:password@localhost/produtos";


// Configurando o banco de dados
const mongoose = require("mongoose");
mongoose.connect(MONGODB_URL, {
    useFindAndModify: false,
    useNewUrlParser: true,
    useUnifiedTopology: true
});

// Mensagem de erro
mongoose.connection.on("error", (err) => {
    console.log("Erro no banco de dados " + err);
});

const produtoSchema = mongoose.Schema({
    media_id: String,
    produtos: [String]
}, {collection: "produtos"});

const Produtos = mongoose.model("Produtos", produtoSchema);


// Habilitando para trabalhar com dados em JSON
app.use(express.json());


// Rotas
app.put("/produtos", (req, res, next) => {
    Produtos.create(req.body).then((produto) => {
        res.send(produto);
    }).catch(next);
});

app.get("/produtos/:media_id", (req, res, next) => {
    Produtos.findOne({media_id: req.params.media_id}).then((produto) => {
        res.send(produto);
    }).catch(next);
});

app.post("/produtos/:media_id", (req, res, next) => {
    Produtos.findOneAndUpdate({media_id: req.params.media_id}, req.body).then(() => {
        Produtos.findOne({media_id: req.params.media_id}).then((produto) => {
            res.send(produto);
        }).catch(next);
    }).catch(next);
});

app.delete("/produtos/:media_id", (req, res, next) => {
    const media_id = req.params.media_id;
    Produtos.findOneAndRemove({media_id: req.params.media_id}).then((produto) => {
        res.send(produto);
    }).catch(next);
});

app.use((err, req, res, next) => {
    res.status(422).send({erro: err.message});
});

app.listen(PORT);
